Develop

misicakovaa opened the pull request:
**Changes**:
- Localizable
- CodingKeys
- Folder organization
- Generics (titled section)
- Combine networking
- Image extension

misicakovaa committed:
VStack alignment instead of HStack & Spacer, Rockets manager - StateObject

misicakovaa committed:
Localizable added

misicakovaa committed:
CodingKeys added

misicakovaa committed:
Folder reorganization - by features, Rockets manager separated

misicakovaa committed:
Generics added (TitledSection)

misicakovaa committed:
Image extension added

misicakovaa committed:
StageRow added

misicakovaa committed:
Combine Rockets manager

misicakovaa committed:
Rockets fetcher added (subscriber)

misicakovaa committed:
New networking

Finemas commented on Rocket/Rocket/LaunchRocket/DeviceMotionHandler.swift:
```diff
@@ -11,9 +11,9 @@ import CoreMotion
+    private var pitch = 0.0
+    private let motionHandler = CMMotionManager()
     
     @Published var deviceRotationChanged = false
```
DeviceMotionHandler by spíš možná měl vracet ten **pitch** a tu logiku, kdy je deviceRotationChanged nechal na viewModelu, protože je celkem specifická. Co myslíš?

misicakovaa replied:
Yes, to dává smysl

Finemas commented on Rocket/Rocket/Networking/ApiService.swift:
```diff
@@ -0,0 +1,60 @@
+
+struct ApiService {
+    
+    enum URLError: Error {
```
Rozšířit bych tenhle error o víc chyb podle status code např.


```
enum CustomApiError: Error {
    // Bad request (400)
    case badRequest
    // Unauthorized (401)
    case unauthorized
    // Not found (404)
    case notFound
    // Conflict (409)
    case serverFailure
    // Build request error
    case buildRequestError
    // Unexpected nil
    case unexpectedNil
    // Unsupported api
    case unsupportedApi
    // Unknown error
    case unknownError
}
```

Finemas commented on Rocket/Rocket/Networking/ApiService.swift:
```diff
@@ -0,0 +1,60 @@
+        
+        
+        /*
+         // TEST - WITHOUT INTERNET CONNECTION
```
Potřebuješ to ještě?

Finemas commented on Rocket/Rocket/Networking/RocketsManager.swift:
```diff
@@ -0,0 +1,29 @@
+
+class RocketsManager: ObservableObject {
+    
+    let apiService = ApiService()
```
Ještě by dobrý použít Depedency Injection:

https://www.raywenderlich.com/14223279-dependency-injection-tutorial-for-ios-getting-started

Finemas commented on Rocket/Rocket/Networking/RocketsManager.swift:
```diff
@@ -0,0 +1,29 @@
+    
+    func fetchRockets() async throws -> [Rocket] {
+        
+        let url = URL(string: "https://api.spacexdata.com/v3/rockets")!
```
Base url bych dal jako proměnnou toho ApiService.

Finemas commented on Rocket/Rocket/RocketDetail/Components/ParametersSectionView.swift:
```diff
@@ -14,29 +14,29 @@ struct ParametersSectionView: View {
+                    unit: "t")
                     .padding(.trailing)
             }
         }
```
Tohle view je za mě celkem dost specifický asi bych ho dal spíš jako view property rocket detail view.

Finemas commented on Rocket/Rocket/RocketDetail/Components/TitledSection.swift:
```diff
@@ -0,0 +1,37 @@
+    let title: String
+    let content: Content
+    
+    init(title: String, @ViewBuilder content: () -> Content) {
```
Myslím, že tenhle init je zbytečný.

EllaGogo replied:
Ja by som to asi nechala takto. Voláš tam tú closure vnútri initu, takže sa to potom celkom hezky používa zvenku.

```swift
// with init
TitledSection(title: "Hello World") {
    ...
}

// without init
TitledSection(
    title: "Hello World",
    content: { ... }() // does't look SwiftUI-ish 😄 
)
```

Finemas commented on Rocket/Rocket/RocketsList/RocketsListView.swift:
```diff
@@ -0,0 +1,77 @@
+            
+            switch vm.state {
+            case .success(let rockets):
+                
```
Na mě je tu dost prázdných řádek.

EllaGogo commented on Rocket/Rocket/Extensions/StringExtension.swift:
```diff
@@ -0,0 +1,27 @@
+    func formateDateString () -> String {
+        let inputDateFormatter = DateFormatter()
+        
+        inputDateFormatter.dateFormat = "yyyy-MM-dd"
+        let date = inputDateFormatter.date(from: self)!
+        
+        let outputDateFormatter = DateFormatter()
+        outputDateFormatter.dateFormat = "d.M.yyyy"
```
Ak ti tie Stringy chodia z API tak by si sa nikdy nemala spoliehať na to, že ti prídu v správnom formáte. Ak náhodou z API príde niečo divné tak ti crashne appka kvôli force unwrapingu 🙂 Takže toto sa treba určite zbaviť a druhá vec je, že vytváranie DateFormatterov je časovo náročné a nie je dobré ich vytvárať dva krát pri každom volaní tejto funkcie. Dala by som ich ako lazy properties do tejto extension a vo funkcii ich iba použila.

EllaGogo replied:
https://sarunw.com/posts/how-expensive-is-dateformatter/

EllaGogo commented on Rocket/Rocket/Extensions/StringExtension.swift:
```diff
@@ -0,0 +1,27 @@
+    // - Formates dateString from format yyyy-MM-dd to d.M.yyyy
+    // - Returns formatted dateString
+    
+    func formateDateString () -> String {
```
Toto záleží na preferencii programátora, ale obecne sa snažíme funkcie robiť čo najviac prepoužiteľné ako to len ide, takže ja by som pridala možnosť si zvoliť vstupný a výstupný formát, a kľudne im dala default values na `yyyy-MM-dd` a `d.M.yyyy` aby si ich nemusela vždy vyplňovať.

EllaGogo commented on Rocket/Rocket/Networking/ApiService.swift:
```diff
@@ -0,0 +1,60 @@
+        let decoder = JSONDecoder()
+        
+        do {
+            let decodedData = try decoder.decode(T.self, from: data)
+            return decodedData
```
dala by som to na one line `return try decoder.decode(T.self, from: data)` 🙂

EllaGogo commented on Rocket/Rocket/RocketDetail/Components/ParametersSectionView.swift:
```diff
@@ -14,29 +14,29 @@ struct ParametersSectionView: View {
             // Parameters info
             HStack {
-                ParameterView(name: "height", number: Int(height), unit: "m")
+                ParameterView(
+                    name: "height",
+                    number: Int(height),
+                    unit: "m")
```
Bavili sme sa na iOS meetingu, že by sme chceli zachovávať jednotné formátovanie, tak by asi bolo fajn použiť to, čo máme v My2NApp. Myslím, že tam máme nastavený page guide na 120 chars a zalamujeme iba ak to pretečie cez tú čiaru 🙂 
S nastavením toho guidu keď tak bude vedieť pomôcť @Finemas

EllaGogo commented on Rocket/Rocket/RocketsList/RocketsListView.swift:
```diff
@@ -0,0 +1,77 @@
+        .task {
+            await vm.getRockets()
+        }
+        .alert("Error", isPresented: $vm.hasError, presenting: vm.state) { detail in
+            
+            Button("Retry") {
+                
+                Task {
+                    await vm.getRockets()
+                }
+                
+            }
+            
+        } message: { detail in
+            
+            if case let .failed(error) = detail {
+                Text(error.localizedDescription)
+            }
+            
+        }
+        
+    }
```
toto je veľmi zvláštne formátovanie, úprimne ani neviem, čo kam patrí 😄  tá message je k tomu alertu? Asi by som v tomto prípade vôbec nepoužila trailing closures.

EllaGogo commented:
Po dohode s @Finemas som ešte spravila posledné CR. Zároveň som dala approve, aby som vás neblokovala v prípadnom mergi. @misicakovaa ak s nejakými mojimi komentármi nebudeš súhlasiť, tak sa dohodnite s Honzou či ich budete aplikovať, alebo to necháte tak 🙂 

Najväčší problém vidím v tom, že tam nie je dependency injection, ako už spomínal Honza v komentári. Ale inak dobrá práca 👌 už teraz je ten kód na vysokej úrovni 🙂 Držím palce, nech to ide takto hladko aj naďalej 👍

misicakovaa committed:
Changes:
- Dependency injection
- DateFormatter
- DeviceMotionViewModel
- URLError
- ParameterSectionView as view property
